lib/ui/compositing/scene.h

```diff
@@ -44,16 +46,36 @@ class Scene : public RefCountedDartWrappable<Scene> {
         bool checkerboardRasterCacheImages,
         bool checkerboardOffscreenLayers);
 
+  // Returns true if `dispose()` has not been called.
+  bool valid();
+
   void RasterizeToImage(uint32_t width,
                         uint32_t height,
+                        float pixel_ratio,
                         Dart_Handle raw_image_handle);
 
-  // This is a shared_ptr to support flattening the layer tree from the UI
-  // thread onto the raster thread - allowing access to the texture registry
-  // required to render TextureLayers.
+  std::unique_ptr<LayerTree> BuildLayerTree(uint32_t width,
+                                            uint32_t height,
+                                            float pixel_ratio);
+
+  flutter::LayerTree::Config layer_tree_config_;
+
+  // Fetches the pixel ratio from view 0, or if the window doesn't exist,
+  // fallback to 2.0f.
+  //
+  // The pixel ratio is used in toImage() and toImageSync(), and its only effect
+  // is to calculate the device's physical dimension, which is used by some
+  // physical shapes (see PhysicalShapeLayer).
+  //
+  // Physical shapes have been deprecated and should be removed soon. This
+  // method aims to keep the legacy behavior in single-window Flutter, which
+  // feeds the toImage and toImageSync with the pixel ratio of the only window.
   //
-  // No longer valid after calling `takeLayerTree`.
-  std::shared_ptr<flutter::LayerTree> layer_tree_;
+  // TODO(dkwingsmt): If PhysicalShapeLayer has been removed as well as
+  // {Preroll,Paint}Context.frame_device_pixel_ratio, remove this method and its
+  // related logic.
+  // https://github.com/flutter/flutter/issues/125720
+  static float defaultViewPixelRatio();
```

Comment:
Can we give this a name that clearly indicated that this is legacy behavior and new code should not depend on this? Maybe we can replace `default` with `legacy` or something like that... The "default" should be to _not_ assume a specific view or DPR.